.teamcity/builds/Build.kt

```diff
@@ -0,0 +1,153 @@
+package builds
+
+import jetbrains.buildServer.configs.kotlin.BuildType
+import jetbrains.buildServer.configs.kotlin.Project
+import jetbrains.buildServer.configs.kotlin.buildFeatures.notifications
+import jetbrains.buildServer.configs.kotlin.sequential
+import jetbrains.buildServer.configs.kotlin.toId
+
+class Build(
+    name: String,
+    forPullRequests: Boolean,
+    javaVersions: Set<JavaVersion>,
+    scalaVersions: Set<ScalaVersion>,
+    pysparkVersions: Set<PySparkVersion>,
+    neo4jVersions: Set<Neo4jVersion>,
+    forCompatibility: Boolean = false,
+    customizeCompletion: BuildType.() -> Unit = {}
+) :
+    Project(
+        {
+          this.id(name.toId())
+          this.name = name
+
+          val complete = Empty("${name}-complete", "complete")
+
+          val bts = sequential {
+            if (forPullRequests)
+                buildType(WhiteListCheck("${name}-whitelist-check", "white-list check"))
+            if (forPullRequests) dependentBuildType(PRCheck("${name}-pr-check", "pr check"))
+
+            parallel {
+              javaVersions.cartesianProduct(scalaVersions, neo4jVersions).forEach {
+                  (java, scala, neo4j) ->
+                sequential {
+                  val packaging =
```

Comment:
I wonder if it makes sense to re-use this package step based only on `java` and `scala` version for all dependent test builds for each `neo4j` version?  

Maybe we can have a nested loop over neo4j version similarly as we have pythin version nested loop?

Reply:
That was my initial attempt too, but Teamcity complains if a single build is re-used by different contexts and I thought this would be the most straightforward way to make it work (and since we have caching in place it wouldn't be a big loss of time/resources).